libs/pilight/protocols/433.92/unitec.c

```diff
@@ -0,0 +1,245 @@
+}
+
+static void createFooter(void) {
+	unitec_switch->raw[RAW_LENGTH - 2] = MIN_PULSE_LENGTH;
+	unitec_switch->raw[RAW_LENGTH - 1] = OUR_PULSE_DIV;
+	unitec_switch->rawlen = RAW_LENGTH;
+}
+
+static int createCode(const struct JsonNode *code) {
+	const char *id = NULL;
+	int unit = -1;
+	int state = -1;
+	double itmp;
+
+	if(json_find_string(code, "id", &id) != 0) {
+		logprintf(LOG_ERR, "unitec_switch: missing --id argument");
+		return EXIT_FAILURE;
+	}
+
+	if(json_find_number(code, "off", &itmp) == 0)
+		state=0;
+	else if(json_find_number(code, "on", &itmp) == 0)
+		state=1;
+	else {
+		logprintf(LOG_ERR, "unitec_switch: missing state argument (--on or --off)");
+		return EXIT_FAILURE;
+	}
+
+	if(json_find_number(code, "unit", &itmp) == 0) {
+		unit = (int)itmp;
+	} else {
+		logprintf(LOG_ERR, "unitec_switch: missing --unit argument (1..4 or 0 for all)");
+		return EXIT_FAILURE;
+	}
+
+	if(id[0] != 'X' && id[0] != 'x') {
+		logprintf(LOG_ERR, "unitec_switch: id must start with 'X' followed by hex-digits (0..9 and A-F).");
+		return EXIT_FAILURE;
+	}
+	char *end = NULL;
+	int idn = strtol(id+1, &end, 16);
+	if(!end || end==id+1 || *end) {
+		logprintf(LOG_ERR, "unitec_switch: invalid char in id: %s ", end ? end : "?");
+		return EXIT_FAILURE;
+	}
+	if (end -(id+1) > 5) {
+		logprintf(LOG_ERR, "unitec_switch: too long id. Max 5 digits allowed.");
+		return EXIT_FAILURE;
+	}
+
+	if(unit < 0 || unit > 4) {
+		logprintf(LOG_ERR, "unitec_switch: unit not in (1..4 or 0 for all)");
+		return EXIT_FAILURE;
+	}
+
+	createMessage(id, unit, state);
+
+	clearCode();
+	createId(idn);
+	createUnit(unit);
+	createState(state);
+	createFooter();
+
+	return EXIT_SUCCESS;
+}
+
+static void printHelp(void) {
+	printf("\t -t --on\t\t\tsend an on signal\n");
+	printf("\t -f --off\t\t\tsend an off signal\n");
+	printf("\t -u --unit=unit\t\t\tcontrol a device with this unit code (1..4, 0=all)\n");
+	printf("\t -i --id=id\t\t\tcontrol a device with this id (Xhexdigits, e.g. 'X0')\n");
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void unitecSwitchInit(void) {
+
+	protocol_register(&unitec_switch);
+	protocol_set_id(unitec_switch, "unitec_switch");
+	protocol_device_add(unitec_switch, "unitec_switch", "Unitec Switches");
+	unitec_switch->devtype = SWITCH;
+	unitec_switch->hwtype = RF433;
+	unitec_switch->minrawlen = RAW_LENGTH;
+	unitec_switch->maxrawlen = RAW_LENGTH;
+	unitec_switch->maxgaplen = MAX_PULSE_DIV;
+	unitec_switch->mingaplen = MIN_PULSE_DIV;
+
+	options_add(&unitec_switch->options, 't', "on", OPTION_NO_VALUE, DEVICES_STATE, JSON_STRING, NULL, NULL);
+	options_add(&unitec_switch->options, 'f', "off", OPTION_NO_VALUE, DEVICES_STATE, JSON_STRING, NULL, NULL);
+	options_add(&unitec_switch->options, 'u', "unit", OPTION_HAS_VALUE, DEVICES_ID, JSON_NUMBER, NULL, "^[0-5]$");
+	options_add(&unitec_switch->options, 'i', "id", OPTION_HAS_VALUE, DEVICES_ID, JSON_STRING, NULL, "^X[0-9A-Fa-f]+$");
```

Comment:
> The id is a string build from 16 different chars. I could made it a string out of any other number of chars, 8, 19 or 24 or 33 different chars for example. But I choose 16 because this leads to short enough IDs of max. 5 chars.

How can you `choose` a string length. I assume the protocol you are reverse-engineering has a fixed logic, not something you can make various chooses in right?

What you are implementing is just totally different in some aspects than i experiences before, that's why i want to understand what and why...

In the clarus protocol the character is actually an essential part of the raw code being built and is also generated from interpreting the raw code:
https://github.com/pilight/pilight/blob/master/libs/pilight/protocols/433.92/clarus.c#L75-L80
https://github.com/pilight/pilight/blob/master/libs/pilight/protocols/433.92/clarus.c#L141-L157